src/battleSystem/hpGauge.mjs

```diff
@@ -126,76 +350,604 @@ class HPGauge
+			this.drainTimer += this.drainSpeed / Sphere.frameRate;
+
+			if (this.drainTimer < 0.25)
+				this.reading = Math.round(tween(this.oldReading, this.drainTimer, 0.25, this.newReading));
+			else
+				this.reading = this.newReading;
 		}
+
+		if (this.numCombosRunning <= 0 && this._reading == this.newReading)
+			this.damageFadeness = this._damageFadeness + this.drainSpeed / Sphere.frameRate;
+
+		if (this.damage > 0)
+			this.damageShown = Math.min(this.damage, (this.oldReading - this._reading)) * (1.0 - this._damageFadeness);
+		else
+			this.damageShown = 0;
+
+		this.numReservesEmpty = this.reserves.length - this._numReservesDamaged - this._numReservesFilled;
 	}
 
-	update()
+	set damageShown (value)
 	{
-		++this.colorFadeTimer;
-		if (this.colorFadeDuration != 0 && this.colorFadeTimer < this.colorFadeDuration) {
-			this.hpColor.r = tween(this.oldColor.r, this.colorFadeTimer, this.colorFadeDuration, this.newColor.r);
-			this.hpColor.g = tween(this.oldColor.g, this.colorFadeTimer, this.colorFadeDuration, this.newColor.g);
-			this.hpColor.b = tween(this.oldColor.b, this.colorFadeTimer, this.colorFadeDuration, this.newColor.b);
-			this.hpColor.a = tween(this.oldColor.a, this.colorFadeTimer, this.colorFadeDuration, this.newColor.a);
-		} else {
-			this.hpColor = this.newColor;
-			this.colorFadeDuration = 0;
+		if (this._damageShown !== value)
+		{
+			this._damageShown = value;
+			let mainDamage = value % this.sectorSize;
+			this.mainBar.damage = mainDamage * (this.mainBar._width - 2) / this.sectorSize;
+			this.numReservesDamaged = Math.min(Math.floor(value / this.sectorSize), this.numReserves - this._numReservesFilled);
 		}
-		this.fadeness = Math.min(Math.max(this.fadeness + this.fadeSpeed / Sphere.frameRate, 0.0), 1.0);
-		this.drainTimer += this.drainSpeed / Sphere.frameRate;
-		if (this.newReading != this.reading && this.drainTimer < 0.25) {
-			this.reading = Math.round(tween(this.oldReading, this.drainTimer, 0.25, this.newReading));
-		} else {
-			this.reading = this.newReading;
-		}
-		if (this.numCombosRunning <= 0 && this.reading == this.newReading) {
-			this.damageFadeness += this.drainSpeed / Sphere.frameRate;
-			if (this.damageFadeness >= 1.0) {
+	}
+
+	set damageFadeness (value)
+	{
+		if (this._damageFadeness !== value)
+		{
+			if (value >= 1.0)
+			{
 				this.damage = 0;
-				this.damageFadeness = 1.0;
+				this._damageFadeness = 1.0;
+			}
+			else
+			{
+				this._damageFadeness = value;
 			}
 		}
 	}
+
+	set reading (value)
+	{
+		// THE MAGIC
+		if (this._reading !== value) //do nothing if this isn't a change
+		{
+			// 1. set the new value
+			this._reading = value;
+			if (value < 0)
+			{
+				value = 0;
+			}
+			// 2. calculate value for main bar
+			let mainValue = value % this.sectorSize;
+			if (mainValue === 0 && value !== 0)
+			{
+				mainValue = this.sectorSize;
+			}
+
+			// 3. calculate number of full reserves
+			this.numReservesFilled = Math.round((value - mainValue) / this.sectorSize);
```

Comment:
Full reserves = number of complete bars over and above the one currently being worked on, so `Math.round()` seems like the wrong function to use here - I'd expect `floor` (or possibly `ceil`?).  That said the gauge *does* appear to function correctly, so not sure.

Reply:
Math.round is correct here as I subtract mainValue first. mainValue = amount in the main bar. so value / this.sectorSize should actually be an int, the Math.round is for safety as occasionally JS integer division mistakenly coerces to a float and gives unwanted decimal places. this is to e.g. turn 5.0001 back to 5 or 4.9999 to 5.

Reply:
That makes sense, thanks.  My confusion came from the fact that I calculated this value using:
```js
Math.ceil(value / sectorSize - 1)
```
So seeing a Math.round in the same context raised red flags.